src/spikeinterface/core/generate.py

```diff
@@ -722,19 +666,651 @@ def generate_lazy_recording(
+                depolarization_ms=params["depolarization_ms"][u],
+                repolarization_ms=params["repolarization_ms"][u],
+                hyperpolarization_ms=params["hyperpolarization_ms"][u],
+                smooth_ms=params["smooth_ms"][u],
+                dtype=dtype,
+            )
+        
+                
+        alpha = params["alpha"][u]
+        # the espilon avoid enormous factors
+        eps = 1.
+        pow = 1.5
+        # naive formula for spatial decay
+        channel_factors = alpha / (distances[u, :] + eps) ** pow
+        if upsample_factor is not None:
+            for f in range(upsample_factor):
+                templates[u, :, :, f] = wf[f::upsample_factor, np.newaxis] * channel_factors[np.newaxis, :]
+        else:
+            templates[u, :, :] = wf[:, np.newaxis] * channel_factors[np.newaxis, :]
+
+    return templates
+
+
+    
+
+
+## template convolution zone ##
+
+class InjectTemplatesRecording(BaseRecording):
+    """
+    Class for creating a recording based on spike timings and templates.
+    Can be just the templates or can add to an already existing recording.
+
+    Parameters
+    ----------
+    sorting: BaseSorting
+        Sorting object containing all the units and their spike train.
+    templates: np.ndarray[n_units, n_samples, n_channels] or np.ndarray[n_units, n_samples, n_oversampling]
+        Array containing the templates to inject for all the units.
+        Shape can be:
+            * (num_units, num_samples, num_channels): standard case
+            * (num_units, num_samples, num_channels, upsample_factor): case with oversample template to introduce sampling jitter.
+    nbefore: list[int] | int | None
+        Where is the center of the template for each unit?
+        If None, will default to the highest peak.
+    amplitude_factor: list[float] | float | None, default None
+        The amplitude of each spike for each unit.
+        Can be None (no scaling).
+        Can be scalar all spikes have the same factor (certainly useless).
+        Can be a vector with same shape of spike_vector of the sorting.
+    parent_recording: BaseRecording | None
+        The recording over which to add the templates.
+        If None, will default to traces containing all 0.
+    num_samples: list[int] | int | None
+        The number of samples in the recording per segment.
+        You can use int for mono-segment objects.
+
+    Returns
+    -------
+    injected_recording: InjectTemplatesRecording
+        The recording with the templates injected.
+    """
+
+    def __init__(
+        self,
+        sorting: BaseSorting,
+        templates: np.ndarray,
+        nbefore: Union[List[int], int, None] = None,
+        amplitude_factor: Union[List[List[float]], List[float], float, None] = None,
+        parent_recording: Union[BaseRecording, None] = None,
+        num_samples: Union[List[int], None] = None,
```

Comment:
I think `Optional[List[int]]` is better than `Union[List[int], None]`

Reply:
Done